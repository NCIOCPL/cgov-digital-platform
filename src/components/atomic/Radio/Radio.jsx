import React from 'react';
import PropTypes from 'prop-types';
import Utilities from '../../../utilities/utilities';

import './Radio.scss';

const Radio = ({ id, label, className, disabled, value, ...otherProps }) => (
  <div className={`cts-radio ${className}`}>
    <input
      className="cts-radio__input"
      type="radio"
      disabled={disabled}
      aria-disabled={disabled}
      id={id}
      value={value ? value : id}
      {...otherProps}
    />
    <label className="cts-radio__label" htmlFor={id}>
      {label}
    </label>
  </div>
);

Radio.propTypes = {
  id: PropTypes.string,
  label: PropTypes.string.isRequired,
  name: PropTypes.string,
  value: PropTypes.string,
  checked: PropTypes.bool,
  defaultChecked: PropTypes.bool,
  disabled: PropTypes.bool,
  onChange: PropTypes.func,
  className: PropTypes.string,
};

Radio.defaultProps = {
  id: Utilities.uniqueIdForComponent(),
  name: 'radios',
  disabled: false,
  defaultChecked: false,
  className: '',
};

export default Radio;
