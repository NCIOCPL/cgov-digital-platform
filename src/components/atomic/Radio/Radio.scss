.cts-radio {
  box-sizing: border-box;
  padding: 0;
  &__input {
    position: absolute;
    left: -999em;
  }
  &__label {
    cursor: pointer;
    display: inherit;
    font-weight: 500;
    margin-bottom: 0.75em;
    padding-left: 2em;
    position: relative;
		text-indent: -2em;
		
    &::before {
      background: white;
      content: '\a0';
      display: inline-block;
      left: 2px;
      position: relative;
      vertical-align: middle;
      height: 1.25em;
      border-radius: 99rem;
      width: 1.25em;
      box-shadow: 0 0 0 2px #71767a;
      line-height: 1.25em;
      margin-right: 0.75em;
    }
  }
}

.cts-radio__input:checked + .cts-radio__label::before {
  background-color: #005ea2;
  -webkit-box-shadow: 0 0 0 2px #005ea2;
  box-shadow: 0 0 0 2px #005ea2;
}

.cts-radio__input:checked + .cts-radio__label::before {
  -webkit-box-shadow: 0 0 0 2px #005ea2, inset 0 0 0 2px white;
  box-shadow: 0 0 0 2px #005ea2, inset 0 0 0 2px white;
}
.cts-radio__input:focus + .cts-radio__label::before {
  outline: 0.25rem solid #2491ff;
  outline-offset: 0.25rem;
}
.cts-radio__input:disabled + .cts-radio__label::before {
  background: #e6e6e6;
  -webkit-box-shadow: 0 0 0 2px #c9c9c9;
  box-shadow: 0 0 0 2px #c9c9c9;
  cursor: not-allowed;
}

@media print {
  .cts-radio__input:checked + .cts-radio__label::before {
    box-shadow: inset 0 0 0 2px white, inset 0 0 0 1rem #005ea2,
      0 0 0 2px #005ea2;
  }
}
