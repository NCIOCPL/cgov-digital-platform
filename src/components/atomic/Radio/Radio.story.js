import generateStories from '../../../../.storybook/generateStories';
import Radio from './Radio';

generateStories({
  component: { Radio },
  storyName: 'Elements | Radio',
  defaultProps: {
    id: 'storybook-radio',
    label: 'Example Radio',
    value: null,
    defaultChecked: false
  },
  stories: [
    {
      name: 'default'
    }
  ]
});
