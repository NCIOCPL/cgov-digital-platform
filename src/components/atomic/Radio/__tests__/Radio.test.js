import React from 'react';
import { mount } from 'enzyme';

import Radio from '../Radio.jsx';

jest.unmock('../Radio.jsx');

describe('Radio', () => {

  let wrapper = null;
  const label = 'label goes here';


  beforeEach(function() {
    wrapper = mount(<Radio label={label} />);
  });

  it('is defined and renders without error', () => {
    expect(wrapper).toBeDefined();
    expect(wrapper).toMatchSnapshot();
  });

  it('renders the label text correctly', () => {
    const labelElement = wrapper.find('label');
    expect(labelElement.text()).toBe(label);
  });

  it('has an id attribute', () => {
    expect(typeof wrapper.find('input').props().id).toBe('string');
  });

  it('has a matching id and htmlFor attributes', () => {
    const inputId = wrapper.find('input').props().id;
    const labelHtmlFor = wrapper.find('label').props().htmlFor;
    expect(labelHtmlFor).toBe(inputId);
  });

  it('has a default value property', () => {
    const id = wrapper.find('input').props().id;
    const value = wrapper.find('input').props().value;
    expect(value).toBe(id);
  });

  it('accepts a value property', () => {
    const testValue = 'testing';
    wrapper = mount(<Radio value={testValue} label={label} />);
    const inputValue = wrapper.find('input').props().value;
    expect(inputValue).toBe(testValue);
  });

  it('has a default name property', () => {
    const inputName = wrapper.find('input').props().name;
    expect(inputName).toBe('radios');
  });

  it('accepts a name property', () => {
    const testValue = 'testing';
    wrapper = mount(<Radio name={testValue} label={label} />);
    const inputName = wrapper.find('input').props().name;
    expect(inputName).toBe(testValue);
  });

  it('is not checked by default', () => {
    const inputChecked = wrapper.find('input').props().checked;
    expect(inputChecked).toBe(false);
  });

  it('accepts the defaultChecked property', () => {
    wrapper = mount(<Radio defaultChecked label={label} />);
    const inputChecked = wrapper.find('input').props().defaultChecked;
    expect(inputChecked).toBe(true);
  });

  it('is not disabled by default', () => {
    const inputDisabled = wrapper.find('input').props().disabled;
    expect(inputDisabled).toBe(false);
  });

  it('accepts the disabled property', () => {
    wrapper = mount(<Radio disabled label={label} />);
    const inputDisabled = wrapper.find('input').props().disabled;
    expect(inputDisabled).toBe(true);
  });
});