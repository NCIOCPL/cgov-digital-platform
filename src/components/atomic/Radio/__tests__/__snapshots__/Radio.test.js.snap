// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio is defined and renders without error 1`] = `
<Radio
  className=""
  disabled={false}
  id="component-unique-id-0"
  label="label goes here"
  name="radios"
>
  <div
    className="cts-radio "
  >
    <input
      aria-disabled={false}
      className="cts-radio__input"
      disabled={false}
      id="component-unique-id-0"
      name="radios"
      type="radio"
      value="component-unique-id-0"
    />
    <label
      className="cts-radio__label"
      htmlFor="component-unique-id-0"
    >
      label goes here
    </label>
  </div>
</Radio>
`;
